### PR TITLE
Adding T5 to NeMo 2.0

### PR DESCRIPTION
**Pull request by huvunvidia**
# What does this PR do ?

Adding support for T5 in NeMo 2.0. Main files working on:
+ T5 specific:
    - megatron_t5_pretraining.py 
    - pre_training.py (using T5Dataset from mcore)
    - t5.py
+ Others:
    - auto_tokenizer.py: adding inv_vocab() to tokenizer to be compatible with mcore's T5Dataset 
    - lr_scheduler.py: adding missing warmup support

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by ko3n1g — CHANGES_REQUESTED** [review]
There are some unwanted changed in `cicd-main.yml`. Can you please revert these?

Huy: Done

File: nemo/collections/llm/t5/model/t5.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+from nemo.collections.llm import fn
+from nemo.collections.nlp.modules.common.megatron.token_level_encoder_decoder import AttnMaskType
+from nemo.collections.nlp.modules.common.megatron.utils import build_attention_mask_3d
+from nemo.lightning import get_vocab_size, io
+from nemo.lightning.megatron_parallel import MaskedTokenLossReduction
+from nemo.lightning.pytorch.optim import MegatronOptimizerModule, OptimizerModule
+
+HAVE_TE = True
+try:
+    import transformer_engine
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/collections/llm/t5/model/t5.py:20]: remove if not used

File: nemo/lightning/pytorch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -264,21 +264,31 @@ class WarmupAnnealingScheduler(LRSchedulerModule):
 
     def __init__(
         self,
+        warmup_steps: int = 750,
```
Comment (yaoyu-33) [inline nemo/lightning/pytorch/optim/lr_scheduler.py:267]: this will change the behavior of other runs, was it a bug? Do you know if any other model is using this?

**Review by yaoyu-33 — APPROVED** [review]

**Comment by huvunvidia** [comment]
> There are some unwanted changed in `cicd-main.yml`. Can you please revert these?

Done.

**Review by yaoyu-33 — APPROVED** [review]

**Review by pablo-garay — APPROVED** [review]

File: .github/workflows/cicd-main.yml [hunk 3]
```diff
@@ -5222,6 +5222,24 @@ jobs:
...
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_T5_Pretraining') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+        NVTE_FUSED_ATTN=0 NVTE_FLASH_ATTN=0 python tests/collections/llm/megatron_t5_pretraining.py \
+        --devices=2 \
+        --max-steps=3 \
+        --experiment-dir=tests/collections/llm/t5_pretrain_results \
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5235]:
```suggestion
        --experiment-dir=tests/collections/llm/t5_pretrain_results/${{ github.run_id }} \
```

File: .github/workflows/cicd-main.yml [hunk 4]
```diff
@@ -5222,6 +5222,24 @@ jobs:
...
+        --data-path=/home/TestData/nlp/megatron_t5/data/pile_val_small_bert_tokenizer_text_document \
+        --index-mapping-dir=tests/collections/llm/t5_index_mappings
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5237]:
```suggestion
        --index-mapping-dir=tests/collections/llm/t5_index_mappings/${{ github.run_id }}
```

File: .github/workflows/cicd-main.yml [hunk 5]
```diff
@@ -5222,6 +5222,24 @@ jobs:
...
+
+      AFTER_SCRIPT: |
+        rm -rf tests/collections/llm/t5_pretrain_results
+        rm -rf tests/collections/llm/t5_index_mappings
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5241]:
```suggestion
        rm -rf tests/collections/llm/t5_pretrain_results/${{ github.run_id }}
        rm -rf tests/collections/llm/t5_index_mappings/${{ github.run_id }}
```

**Review by ko3n1g — CHANGES_REQUESTED** [review]
Have just two remarks, rest I'm not smart enough to properly review

File: .github/workflows/cicd-main.yml (hunk 5 above)
Comment (huvunvidia) [reply .github/workflows/cicd-main.yml:5242]: Will fix this on a separate PR right after this PR.
This PR is overdue and due to a Github glitch it didn't get merged.

File: .github/workflows/cicd-main.yml (hunk 4 above)
Comment (huvunvidia) [reply .github/workflows/cicd-main.yml:5238]: Will fix this on a separate PR right after this PR.
This PR is overdue and due to a Github glitch it didn't get merged.

File: .github/workflows/cicd-main.yml (hunk 3 above)
Comment (huvunvidia) [reply .github/workflows/cicd-main.yml:5236]: Will fix this on a separate PR right after this PR.
This PR is overdue and due to a Github glitch it didn't get merged.

**Review by yaoyu-33 — APPROVED** [review]
